src/main/java/com/icthh/xm/ms/entity/security/access/DynamicPermissionCheckService.java

```diff
@@ -0,0 +1,154 @@
+        Preconditions.checkArgument(StringUtils.isNotEmpty(basePermission));
+        Preconditions.checkArgument(StringUtils.isNotEmpty(suffix));
+        if (featureContext.isEnabled(this)) {
+            return checkContextPermission(basePermission, suffix);
+        }
+        return assertPermission(basePermission);
+    }
+
+    /**
+     * Checks if user has permission with dynamic key feature permission = basePermission + "." + suffix
+     * @param basePermission - base permission
+     * @param suffix - suffix
+     * @return result result from PermissionCheckService.hasPermission(permission) from assertPermission
+     */
+    public boolean checkContextPermission(String basePermission, String suffix) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(basePermission));
+        Preconditions.checkArgument(StringUtils.isNotEmpty(suffix));
+        final String permission = basePermission + "." + suffix;
+        return assertPermission(permission);
+    }
+
+    /**
+     * Performs object field adjustment if feature is enabled
+     * @param mapper - adjustment implementation
+     * @param <T> - collection element
+     * @return same collection or adjusted one
+     */
+    public <T> Function<List<T>, List<T>> dynamicFunctionFilter(Function<T, T> mapper) {
+        if (FeatureContext.FUNCTION.isEnabled(this)) {
+            return list -> list.stream().map(mapper).collect(Collectors.toList());
+        }
+        return list -> list;
+    }
+
+    /**
+     * Assert permission via permissionCheckService.hasPermission
+     * @param permission
+     * @return
+     */
+    protected boolean assertPermission(final String permission) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(permission));
+        final boolean permitted = assertPermission.apply(permissionCheckService, permission).booleanValue();
+        if (!permitted) {
+            //TODO place correct error here
+            throw new IllegalStateException("Throw correct exception here");
+        }
+        return Boolean.TRUE.booleanValue();
+    }
+
+    /**
+     * Checks if feature tenant-config -> functions -> dynamic enabled
+     * @return true if feature enabled
+     */
+    private boolean isDynamicFunctionPermissionEnabled() {
+        return getTenantConfigBooleanParameterValue(CONFIG_SECTION, DYNAMIC_FUNCTION_PERMISSION_FEATURE)
+            .map(it -> (Boolean)it).orElse(Boolean.FALSE).booleanValue();
+    }
+
+    /**
+     * Checks if feature tenant-config -> stateChange -> dynamic enabled
+     * @return true if feature enabled
+     */
+    private boolean isDynamicChangeStatePermissionEnabled() {
+        throw new UnsupportedOperationException(this + " Not implementer");
+        /*return getTenantConfigBooleanParameterValue("xxx", "")
+            .map(it -> (Boolean)it).orElse(Boolean.FALSE).booleanValue();*/
+    }
+
+    // TODO should be in Commons.TenantConfigService as utility method
```

Comment:
I see no reason to postpone
let's not shift work to others and create tech dept
let's do it right away now

Reply:
will be implemented on phase 2 after discussion and design